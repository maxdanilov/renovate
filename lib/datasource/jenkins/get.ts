import { logger } from '../../logger';
import { ExternalHostError } from '../../types/errors/external-host-error';
import { Http } from '../../util/http';
import { Release, ReleaseResult } from '../common';
import { id } from './common';

const http = new Http(id);
const cacheTimeMin = 60;

const packageInfoUrl =
  'https://updates.jenkins.io/current/update-center.actual.json';

let lastSync = new Date('2000-01-01');
let packageReleases: Record<string, ReleaseResult> = Object.create(null);

// Note: use only for tests
export function resetCache(): void {
  lastSync = new Date('2000-01-01');
  packageReleases = Object.create(null);
}

function processPlugin(name: string, data: Record<string, any>): void {
  const pluginReleases: Release[] = [data?.previousVersion, data?.version]
    .filter((x) => x !== null && x !== undefined)
    .map((x) => {
      return { version: x };
    });

  packageReleases[name] = {
    name,
    releases: pluginReleases,
    sourceUrl: data?.scm,
  };
}

async function updateJenkinsPluginVersions(): Promise<void> {
  const options = {
    headers: {
      'Accept-Encoding': 'gzip, deflate, br',
    },
  };
  let body: string;

  try {
    logger.debug('Jenkins: Fetching Jenkins plugins versions');
    let startTime = Date.now();

    body = JSON.parse((await http.get(packageInfoUrl, options)).body);
    let durationMs = Math.round(Date.now() - startTime);
    logger.debug({ durationMs }, 'Jenkins: Fetched Jenkins plugins versions');

    startTime = Date.now();
    durationMs = Math.round(Date.now() - startTime);
    logger.debug({ durationMs }, 'Jenkins: Parsed Jenkins plugins list');
  } catch (err) /* istanbul ignore next */ {
    packageReleases = Object.create(null);
    throw new ExternalHostError(
      new Error('Jenkins: Fetch error - need to reset cache')
    );
  }

  if (body) {
    const pluginsData: Record<string, any> = ((body as unknown) as Record<
      string,
      any
    >).plugins;

    for (const pluginName in pluginsData) {
      if (Object.prototype.hasOwnProperty.call(pluginsData, pluginName)) {
        processPlugin(pluginName, pluginsData[pluginName]);
      }
    }
  }

  lastSync = new Date();
}

function isDataStale(): boolean {
  const minutesElapsed = Math.floor(
    (new Date().getTime() - lastSync.getTime()) / (60 * 1000)
  );
  return minutesElapsed >= cacheTimeMin;
}

let _updateJenkinsPluginVersions: Promise<void> | undefined;

async function syncVersions(): Promise<void> {
  if (isDataStale()) {
    _updateJenkinsPluginVersions =
      // eslint-disable-next-line @typescript-eslint/no-misused-promises
      _updateJenkinsPluginVersions || updateJenkinsPluginVersions();
    await _updateJenkinsPluginVersions;
    _updateJenkinsPluginVersions = null;
  }
}

export async function getJenkinsPluginDependency(
  lookupName: string
): Promise<ReleaseResult | null> {
  logger.debug(`getJenkinsDependency(${lookupName})`);
  await syncVersions();
  if (!packageReleases[lookupName]) {
    return null;
  }

  return packageReleases[lookupName];
}
