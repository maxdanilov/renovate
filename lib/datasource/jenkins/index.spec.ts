import { getPkgReleases } from '..';
import * as httpMock from '../../../test/httpMock';
import * as versioning from '../../versioning/docker';
import jenkinsVersions from './__fixtures__/update-center.actual.json';
import { resetCache } from './get';
import * as jenkins from '.';

describe('datasource/jenkins', () => {
  describe('getReleases', () => {
    const SKIP_CACHE = process.env.RENOVATE_SKIP_CACHE;

    const params = {
      versioning: versioning.id,
      datasource: jenkins.id,
      depName: 'email-ext',
      registryUrls: ['https://updates.jenkins.io/'],
    };

    beforeEach(() => {
      resetCache();
      httpMock.setup();
      process.env.RENOVATE_SKIP_CACHE = 'true';
      jest.resetAllMocks();
    });

    afterEach(() => {
      httpMock.reset();
      process.env.RENOVATE_SKIP_CACHE = SKIP_CACHE;
    });

    it('returns null for a package miss', async () => {
      const newparams = { ...params };
      newparams.depName = 'non-existing';

      httpMock
        .scope('https://updates.jenkins.io')
        .get('/current/update-center.actual.json')
        .reply(200, jenkinsVersions);

      expect(await getPkgReleases(newparams)).toBeNull();
    });

    it('returns package releases for a hit', async () => {
      httpMock
        .scope('https://updates.jenkins.io')
        .get('/current/update-center.actual.json')
        .reply(200, jenkinsVersions);

      const res = await getPkgReleases(params);
      expect(res.releases).toHaveLength(2);
      expect(res).toMatchSnapshot();

      expect(res.sourceUrl).toBe(
        'https://github.com/jenkinsci/email-ext-plugin'
      );
      expect(res.name).toBe('email-ext');

      expect(
        res.releases.find((release) => release.version === '2.69')
      ).toBeDefined();
      expect(
        res.releases.find((release) => release.version === '2.10')
      ).toBeUndefined();
    });

    it('returns null empty response', async () => {
      httpMock
        .scope('https://updates.jenkins.io')
        .get('/current/update-center.actual.json')
        .reply(200, '{}');
      expect(await getPkgReleases(params)).toBeNull();
    });
  });
});
