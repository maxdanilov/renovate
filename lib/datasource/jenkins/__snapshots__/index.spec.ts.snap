// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datasource/jenkins getReleases returns package releases for a hit 1`] = `
Object {
  "name": "email-ext",
  "releases": Array [
    Object {
      "version": "2.67",
    },
    Object {
      "version": "2.69",
    },
  ],
  "sourceUrl": "https://github.com/jenkinsci/email-ext-plugin",
}
`;
